File: spark-study-scala/src/main/java/com/zhouq/spark/sql/RDD2DataFrameReflection.scala
package com.zhouq.spark.sql

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.{DataFrame, Row, SQLContext}
import org.apache.spark.{SparkConf, SparkContext}

/**
  *
  * Create by zhouq on 2019/9/8
  * 使用反射的方式将RDD 映射成 DataFrame
  *
  *
  * 如果要用 scala 开发 spark 程序，
  * 然后在其中使用基于反射的RDD 到DataFrame 的转换，就必须得用 object extends App 的方式
  * 不能用 def main 的方式来启动，不然会报 no typetag for 。。。 class 的错误
  */
object RDD2DataFrameReflection extends App {
  val conf: SparkConf = new SparkConf().setAppName("RDD2DataFrameReflection").setMaster("local")
  val sc = new SparkContext(conf)
  val sqlContext = new SQLContext(sc)

  case class Student(id: Int, name: String, age: Int)

  // 在Scala 中使用反射方式，进行RDD 到 DataFrame 的转换，需要手动导入一个 隐式转换，才可以使用它的 toDF() 方法
  import sqlContext.implicits._

  val studentsDF = sc.textFile("H:\\bigdatatest\\spark\\students.txt", 1)
    .map(line => line.split(","))
    .map(arr => Student(arr(0).toInt, arr(1).trim, arr(2).toInt))
    .toDF()

  studentsDF.registerTempTable("students")

  val teenagerDF: DataFrame = sqlContext.sql("select * from students where age <= 18")

  val teenagerRDD: RDD[Row] = teenagerDF.rdd

  teenagerRDD.map(row => Student(row.getInt(0), row.getString(1), row.getInt(2)))
    .collect()
    .foreach(println)

  // scala 中，row中的列的顺序是按照我们期望的顺序来排列的。
  // 也可以使用 row.getAs() 方法，获取指定的列，比在Java 中更丰富
  teenagerRDD.map(row => Student(row.getAs[Int]("id"), row.getAs[String]("name"), row.getAs[Int]("age")))
    .collect()
    .foreach(println)

  // 还有种方式 使用 row.getValuesMap，获取指定的列的数据
  teenagerRDD.map(row => {
    val map: Map[String, Any] = row.getValuesMap[Any](Array("id", "name", "age"))

    Student(map("id").toString.toInt, map("name").toString, map("age").toString.toInt)
  }).collect().foreach(println)

  sc.stop()

}

File: spark-study-java/src/main/java/com/zhouq/spark/sql/RDD2DataFrameReflection.java
package com.zhouq.spark.sql;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SQLContext;

import java.io.Serializable;
import java.util.List;

/**
 * 使用反射的方式将RDD 映射成 DataFrame
 * Create by zhouq on 2019/9/7
 */
public class RDD2DataFrameReflection {

    public static void main(String[] args) {
        SparkConf conf = new SparkConf().setMaster("local").setAppName("RDD2DataFrameReflection");

        JavaSparkContext sc = new JavaSparkContext(conf);

        SQLContext sqlContext = new SQLContext(sc);

        JavaRDD<String> lines = sc.textFile("H:\\bigdatatest\\spark\\students.txt");

        JavaRDD<Student> students = lines.map(new Function<String, Student>() {
            @Override
            public Student call(String line) throws Exception {
                String[] lineSplited = line.split(",");
                Student student = new Student();
                student.setId(Integer.valueOf(lineSplited[0].trim()));
                student.setName(lineSplited[1].trim());
                student.setAge(Integer.valueOf(lineSplited[2].trim()));
                return student;
            }
        });

        //使用反射方式，将RDD 转换为 DataFreme
        //将Student.class 传入进去，其实就是用反射的方式来创建 DataFrame
        //因为 Student.class 本身就是反射的一个应用
        //然后底层还是得通过 Student Class 进行反射，来获取 其中的field
        DataFrame studentDF = sqlContext.createDataFrame(students, Student.class);

        // 拿到 一个DataFrame 以后，注册为一个临时表，然后针对其中的数据执行 SQL 语句
        studentDF.registerTempTable("students");

        // 针对 students 临时表，执行sql 语句，查询年龄小于等于 18岁的学生。就是 teenageer
        DataFrame teenageerDF = sqlContext.sql("select * from students where  age <= 18");

        // 将查询出来的 DataFrame 再次转化为 RDD
        JavaRDD teenagerRDD = teenageerDF.toJavaRDD();

        // 将RDD中的数据，进行映射成Student

        /**
         * 注意： 在Java 中，Row 中的列的数据可能会跟我们期望的顺序不一致
         * 可以使用row.getAs(fieldName) 来获取值
         *
         */
        JavaRDD<Student> teenagerStudentRDD = teenagerRDD.map(new Function<Row, Student>() {
            @Override
            public Student call(Row row) throws Exception {
                System.out.println(row);
                Student stu = new Student();
                stu.setId(row.getInt(0));
                stu.setName(row.getString(2));
                stu.setAge(row.getInt(1));
                return stu;
            }
        });

        // 把数据 collect 回来，打印一下
        List<Student> studentList = teenagerStudentRDD.collect();

        for (Student student : studentList) {
            System.out.println(student);
        }

        sc.stop();
    }


    /**
     * Java Bean
     */
    public static class Student implements Serializable {
        private int id;
        private String name;
        private int age;

        public Student() {

        }

        public Student(int id, String name, int age) {
            this.id = id;
            this.name = name;
            this.age = age;
        }

        public int getId() {
            return id;
        }

        public void setId(int id) {
            this.id = id;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public int getAge() {
            return age;
        }

        public void setAge(int age) {
            this.age = age;
        }

        @Override
        public String toString() {
            return "Student{" +
                    "id=" + id +
                    ", name='" + name + '\'' +
                    ", age=" + age +
                    '}';
        }
    }
}
